### PR TITLE
WIP Cleaning Frankenstein 

### PR DESCRIPTION
**Pull request by totto82**
Should not be merged yet. See TODO (1) 

This PR
1) Removes most of the Eigen stuff. (Only minor stuff in getConvergenc() remains)
2) Add a wellModelMatrixAdapter that applies the effect of the wells to the matrix-vector product supplied to the solver "Ax" = Ax - Binv(D)Cx
3) Rewrites the updateState. The new updateState is simpler(TM) and improves the convergence for SPE9 and norne. (Not yet tested on Model2, i.e. some tuning may still be needed) 

TODO 
1) Copy the MatrixBlock class in NewtonIterationBlackoilInterleaved.cpp to overwrite the invert() in istl. to avoid problems with singular local matrices in iLU. (Norne will fail at one point without it) 
2) Make use of param_ in the solver
3) Remove the last remaining Eigen dependencies in getConvergence()

### CODE REVIEW

File: opm/autodiff/BlackoilModelEbos.hpp [hunk 1]
```diff
@@ -426,309 +381,209 @@ namespace Opm {
...
-            Operator opA(ebosJac);
+            typedef WellModelMatrixAdapter<Mat,BVector,BVector, ThisType> Operator;
+            Operator opA(ebosJac, const_cast< ThisType&  > (*this));
             const double relax = 0.9;
             typedef Dune::SeqILU0<Mat, BVector, BVector> SeqPreconditioner;
             SeqPreconditioner precond(opA.getmat(), relax);
             Dune::SeqScalarProduct<BVector> sp;
 
+            // apply well residual to the residual.
+            wellModel().apply(ebosResid);
```
Comment (andlaus) [inline opm/autodiff/BlackoilModelEbos.hpp]: is the influence of the well model in the residual's Jacobian not considered? that might explain some of the convergence issues we were observing (some of those of the master branch, too).

File: opm/autodiff/BlackoilModelEbos.hpp [hunk 2]
```diff
@@ -426,309 +381,209 @@ namespace Opm {
...
 
         /// Apply an update to the primary variables, chopped if appropriate.
         /// \param[in]      dx                updates to apply to primary variables
         /// \param[in, out] reservoir_state   reservoir state variables
         /// \param[in, out] well_state        well state variables
-        void updateState(const V& dx,
-                         ReservoirState& reservoir_state,
-                         WellState& well_state)
+        void updateState(const BVector& dx,
+                         ReservoirState& reservoir_state)
```
Comment (andlaus) [inline opm/autodiff/BlackoilModelEbos.hpp]: honestly, I don't really understand what this method does. probably it deserves to die...

(if you think the same, add it to the *end* of your TODO list; if you're busy with more important things, I'll have a go at it in the medium term :))

**Comment by andlaus** [comment]
nice! after skimming over it I haven't seen any obvious big red flags. (I skipped the StandardWellsDense class, though.)

File: opm/autodiff/BlackoilModelEbos.hpp (hunk 2 above)
Comment (totto82) [inline opm/autodiff/BlackoilModelEbos.hpp]: This method updates the reservoir_state and is definitely needed. It does the Appelyard chopping and without it the solution would not converge. I implemented a much  simpler version of it when I refactored it. Moreover this new version improves the convergence significantly. Only tested on SPE1, SPE9 and Norne, so some more tuning may be needed.

File: opm/autodiff/BlackoilModelEbos.hpp (hunk 1 above)
Comment (totto82) [inline opm/autodiff/BlackoilModelEbos.hpp]: Yes it is. It happens in the WellModelMatrixAdapter. By putting it in the MatrixAdapter you avoid matrix-matrix multiplications and copying of matrices. Again the speed up compared with the earlier version Frankenstein is significant.

File: opm/autodiff/BlackoilModelEbos.hpp [hunk 3]
```diff
@@ -221,24 +206,29 @@ namespace Opm {
...
             IterationReport iter_report = assemble(timer, iteration, reservoir_state, well_state);
-            residual_norms_history_.push_back(computeResidualNorms());
-            const bool converged = getConvergence(timer, iteration);
-            const bool must_solve = (iteration < nonlinear_solver.minIter()) || (!converged);
+            std::vector<double> residual_norms;
+            const bool converged = getConvergence(timer, iteration,residual_norms);
+            residual_norms_history_.push_back(residual_norms);
+            const bool must_solve = true;
+            Dune::InverseOperatorResult result;
             if (must_solve) {
```
Comment (totto82) [inline opm/autodiff/BlackoilModelEbos.hpp]: Ops. The current version solves the linear system even if the solution is converged. I will push a fix.

**Comment by totto82** [comment]
With the latest PRs also flow runs.

File: opm/autodiff/BlackoilModelEbos.hpp (hunk 2 above)
Comment (andlaus) [inline opm/autodiff/BlackoilModelEbos.hpp]: ok. sounds like stuff for another pr. thanks for the info.

**Comment by andlaus** [comment]
merge it once you consider it to be ready...

**Comment by totto82** [comment]
Now  THP is also supported. 

This shouldn't be merged before the problems with ilu is solved. (The TODO#1)

**Comment by totto82** [comment]
With this Model 2 does not seg-fault any longer.

**Comment by blattms** [comment]
Just a humble comment concerning matrices for wells (commit a966344): You should put "drop using matrix representations for wells" on the todo list. (As this is no neither matlab  nor Eigen)

With so many zero rows setting them up and even iterating over *all* rows seems like a waste of memory bandwidth. If you need the abstraction implement Dune::LinearOperator directly on the well informations and the data values. Your local_invert could be implemented to only iterate in far less steps with this approach. 

But as always I might be missing something.

**Comment by totto82** [comment]
@blattms A good point, still the total assembly time including the local_invert of the D matrix + the local solve of the well equation take less than 0.5% for norne with flow_ebos. Also the time spent in the schur complement done in the MatrixWellAdapter seems to be insignificant. In contrast, flow spends in the order of 10-15% on well related operations.

**Comment by andlaus** [comment]
this PR should now be included in #818. Closing it for that reason. (BTW: if you're happy with #818, please press green.)

**Pull request closed without merging**